### ISSUE TITLE
MiLoPyP particle refined localization training

### ISSUE THREAD

**Issue by PeterVanF** [nextpyp/cet_pick#6]
Hi,
I am using the MiLoPyp for the rubisco particle picking.  I directly used the particle position coordinates and cryocare-denoised ( around 7 A/pixel) tomograms for the MiLoPyP particle refined localization module training. However, I faced the following problems: Loss value became NaN. Do you guys have some suggestions? 

(MiLoPYP2) petervan@GLaDOS:~/Software/cet_pick/cet_pick$ python main.py semi --down_ratio 2 --num_epochs 10 --bbox 22 --exp_id rubiscosample_refinement --dataset semi --arch unet_5 --save_all --debug 4  --thresh 0.85 --cr_weight 0.1 --temp 0.07 --tau 0.01 --lr 5e-4 --train_img_txt rubiscotest_img.txt --train_coord_txt rubisco_training_coordinates.txt  --K 1000 --compress --order xyz  --last_k 3 --gpus 0 --val_interval 1 --val_img_txt rubiscotest_img.txt --val_coord_txt rubisco_training_coordinates.txt --contrastive
Using tensorboardX
Training chunk_sizes: [1]
The output will be saved to  /home/petervan/Software/cet_pick/cet_pick/exp/semi/rubiscosample_refinement
heads {'hm': 1, 'proj': 32}
Creating model...
Setting up data...
Loaded val 6 samples
Loaded train 322 samples
Starting training...
semi/rubiscosample_refinement |                                | train: [1][0/322]|Tot: 0:00:01 |ETA: 0:00:00 |loss 0.1970 |hm_loss 0.1747 |cr_loss 0.2232 |consis_loss 0.0000 |Datasemi/rubiscosample_refinement |                                | train: [1][1/322]|Tot: 0:00:02 |ETA: 0:08:19 |loss 0.2081 |hm_loss 0.1746 |cr_loss 0.3346 |consis_loss 0.0000 |Datasemi/rubiscosample_refinement |                                | train: [1][2/322]|Tot: 0:00:03 |ETA: 0:06:42 |loss nan |hm_loss nan |cr_loss nan |consis_loss 0.0000 |Data 0.025s(0semi/rubiscosample_refinement |                                | train: [1][3/322]|Tot: 0:00:04 |ETA: 0:06:09 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.025s(0.02semi/rubiscosample_refinement |                                | train: [1][4/322]|Tot: 0:00:05 |ETA: 0:05:53 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.026s(0.02semi/rubiscosample_refinement |                                | train: [1][5/322]|Tot: 0:00:06 |ETA: 0:05:41 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.026s(0.02semi/rubiscosample_refinement |                                | train: [1][6/322]|Tot: 0:00:07 |ETA: 0:05:33 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.025s(0.02semi/rubiscosample_refinement |                                | train: [1][7/322]|Tot: 0:00:08 |ETA: 0:05:28 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.036s(0.02semi/rubiscosample_refinement |                                | train: [1][8/322]|Tot: 0:00:09 |ETA: 0:05:25 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.036s(0.02semi/rubiscosample_refinement |                                | train: [1][9/322]|Tot: 0:00:10 |ETA: 0:05:35 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.025s(0.02semi/rubiscosample_refinement |#                               | train: [1][10/322]|Tot: 0:00:12 |ETA: 0:05:40 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.024s(0.0semi/rubiscosample_refinement |#                               | train: [1][11/322]|Tot: 0:00:13 |ETA: 0:05:28 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.025s(0.0semi/rubiscosample_refinement |#                               | train: [1][12/322]|Tot: 0:00:14 |ETA: 0:05:35 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.026s(0.0semi/rubiscosample_refinement |#                               | train: [1][13/322]|Tot: 0:00:15 |ETA: 0:05:43 |loss nan |hm_loss nan |cr_loss nan |consis_loss nan |Data 0.026s(0.0

Best,
PeterVan

**Comment by huangqinwen** [comment]
this is odd..typically this happens when the particle coordinates are off. Is the tomogram in x-y-z order and the particle coordinates in x y z order as well? A possible way to test it out is to use the same files for both training and validation and set --val_intervals to 1  and check the debug folder for *_gt_hm.png files

**Comment by huangqinwen** [comment]
will be adding a function to visually check if input tomogram and coordinates match

**Comment by PeterVanF** [comment]
Hi Qinwen,
the tomogram and particle coordinates are both in x y z order. I also used these for the cryolo training. I also set the  same file for both training and validation and set --val_intervals to 1. But I didn't find  *_gt_hm.png files in my debug folder.

**Comment by huangqinwen** [comment]
Is there anything in the debug folder?

**Comment by huangqinwen** [comment]
Just to quickly test if it’s the input file problem - for the header row
can you change x_coord to y_coord and y_coord to x_coord? And rerun to see
if the nan error still exists?

**Comment by PeterVanF** [comment]
Hi Qinwen,
I changed the x_coord to y_coord and y_coord to x_coord on the header row. it still exist the nan value in the loss. I haven't seen anything in the debug folder.  Another question regarding the bb_boxize. Additionally, I have a question about bb_boxize. I used the particle diameter in voxels calculated from my tomogram. Is it appropriate?

**Comment by huangqinwen** [comment]
we updated the code and hopefully it will solve the problem/or give us a clear picture on why NaNs are happening. can you retry using the following command: 
```python main.py semi --down_ratio 2 --num_epochs 10 --bbox 22 --exp_id rubiscosample_refinement --dataset semi --arch unet_5 --save_all --debug 4 --thresh 0.85 --cr_weight 0.1 --temp 0.07 --tau 0.01 --lr 5e-4 --train_img_txt rubiscotest_img.txt --train_coord_txt rubisco_training_coordinates.txt --K 1000 --compress --order xyz --last_k 3 --gpus 0 --val_intervals 1 --val_img_txt rubiscotest_img.txt --val_coord_txt rubisco_training_coordinates.txt --contrastive --translation_ratio 0.2```

**Comment by PeterVanF** [comment]
OK, I will try it. Thank you Qinwen!
